### PR TITLE
fix: improve v2 database migration robustness with schema validation and self-healing

### PR DESCRIPTION
**Pull request by tphakala**
## Summary

- Fix silent error in `cleanupLegacySchemaContamination()` that permanently stalls v2 database migration (Sentry BIRDNET-GO-RZ: 38 events across 4 systems)
- Add `validateV2SchemaIntegrity()` using GORM schema parser to detect legacy contamination before AutoMigrate runs
- Add self-healing for empty corrupt v2 databases with crash-loop prevention
- Extract `v2Entities()` as single source of truth for v2 entity list

## Details

**Root cause:** `cleanupLegacySchemaContamination()` returned void, silently ignoring `ALTER TABLE DROP COLUMN` failures. This left orphaned columns in v2 tables, causing `AutoMigrate` to fail with "Cannot add NOT NULL column with default value NULL" on every restart — permanently blocking migration.

**Three layers of defense:**
1. **Cleanup error propagation** — DROP COLUMN failures now return `ErrV2SchemaCorrupted`. Falls back to dropping empty tables when DROP COLUMN is unavailable (SQLite < 3.35.0)
2. **Schema integrity validation** — Compares actual SQLite columns against GORM struct definitions. Drops empty contaminated tables, returns error for populated ones
3. **Self-healing** — When `Initialize()` fails with `ErrV2SchemaCorrupted` and the database has no user data, automatically deletes and recreates. Safety: only triggers on schema corruption (not transient errors), checks detections/alert_rules/reviews/migration_state, writes `.v2_schema_reset` marker to prevent crash loops

## Related Issues

Relates to #2672 (incomplete database v2 migration)

## Test Plan

- [x] 5 new tests in `manager_test.go` covering cleanup, validation, and edge cases
- [x] All 73 existing v2 datastore tests pass with `-race`
- [x] golangci-lint: 0 issues (29 active linters)
- [x] Local CodeRabbit review: 2 findings fixed (COUNT error check, os.Stat handling)
- [x] Gemini review: 3 findings — 1 false positive (Go 1.25 WaitGroup.Go), 2 fixed (main file removal error, legacy table name)
- [x] Manual code review: no critical/high issues remaining

## Summary by CodeRabbit

* **Bug Fixes**
  * Detects and safely recovers from corrupted local database schemas, avoiding repeated recovery loops.
  * Prevents unsafe deletions by verifying database contents before reset.

* **Improvements**
  * Stronger startup validation and self-healing for local databases to improve reliability and stability.
  * Better handling when database files are recreated during recovery.

* **Tests**
  * Added coverage validating contamination detection, cleanup, and recovery behaviors.

### CODE REVIEW

File: internal/analysis/database_service.go [hunk 1]
```diff
@@ -307,6 +314,166 @@ func (d *DatabaseService) closeDataStore(log logger.Logger) {
...
+		return false
+	}
+	defer func() {
+		if sqlDB, dbErr := db.DB(); dbErr == nil {
+			_ = sqlDB.Close()
+		}
+	}()
+
+	// Check user-data tables: if ANY has rows, the database is not safe to delete.
+	userTables := []string{"detections", "alert_rules", "detection_reviews"}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/analysis/database_service.go]: Fixed in d9cb7acc — added `dynamic_thresholds` to the user-data safety check.

File: internal/analysis/database_service.go [hunk 2]
```diff
@@ -307,6 +314,166 @@ func (d *DatabaseService) closeDataStore(log logger.Logger) {
...
+	var migrationTableName string
+	if err := db.Raw("SELECT name FROM sqlite_master WHERE type='table' AND name IN ('migration_states', 'migration_state') LIMIT 1").Scan(&migrationTableName).Error; err == nil && migrationTableName != "" {
+		var state string
+		if err := db.Raw("SELECT state FROM `" + migrationTableName + "` WHERE id = 1").Scan(&state).Error; err == nil {
+			unsafeStates := map[string]bool{
+				"migrating":  true,
+				"completed":  true,
+				"dual_write": true,
+				"validating": true,
+			}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/analysis/database_service.go:467]: Fixed in d9cb7acc — now using `entities.MigrationStatus*` constants instead of hardcoded strings.

File: internal/analysis/database_service.go [hunk 3]
```diff
@@ -307,6 +314,166 @@ func (d *DatabaseService) closeDataStore(log logger.Logger) {
...
+			if unsafeStates[state] {
+				log.Warn("v2 database has active migration state, not safe to delete",
+					logger.String("migration_state", state))
+				return false
+			}
+		}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/analysis/database_service.go:473]: Partially fixed in d9cb7acc — added `cutover` and `paused` to unsafe states. Did not add `initializing` or `failed` since those are exactly the states where self-healing should be allowed to proceed (initializing = the state we're fixing, failed = safe to retry).

File: internal/analysis/database_migration.go [hunk 4]
```diff
@@ -250,8 +251,23 @@ func initializeMigrationInfrastructure(settings *conf.Settings, ds datastore.Int
 			Build()
 	}
 
-	// Initialize the v2 database schema
-	if err := v2Manager.Initialize(); err != nil {
+	// Initialize the v2 database schema with self-healing for empty corrupt databases.
+	v2Path := datastoreV2.V2MigrationPathFromConfigured(settings.Output.SQLite.Path)
+	if err := initializeV2WithSelfHealing(v2Manager, v2Path, log); err != nil {
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/analysis/database_migration.go:256]: Valid design observation. The current code handles this via the os.Stat check with explicit non-IsNotExist error handling (commit 58da16a0). Returning `(bool, error)` would be cleaner but the current approach works correctly. Noting as a potential future refinement.

File: internal/analysis/database_service.go [hunk 5]
```diff
@@ -307,6 +314,168 @@ func (d *DatabaseService) closeDataStore(log logger.Logger) {
...
+// initializeV2WithSelfHealing wraps v2 Manager.Initialize() with automatic
+// recovery for empty, schema-corrupted databases. When Initialize() returns
+// ErrV2SchemaCorrupted and the database contains no user data, the database
+// file (plus WAL/SHM) is deleted so the caller can recreate the manager and
+// retry. A marker file prevents infinite crash loops.
+//
+// Returns nil when the database was successfully deleted (caller must recreate
+// the manager and call Initialize again), or the original/wrapped error when
+// self-healing is not possible.
+func initializeV2WithSelfHealing(manager datastoreV2.Manager, v2Path string, log logger.Logger) error {
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/analysis/database_service.go:332]: Acknowledged — see reply on the caller side. Current approach is correct if slightly indirect.

File: internal/analysis/database_service.go [hunk 6]
```diff
@@ -307,6 +314,168 @@ func (d *DatabaseService) closeDataStore(log logger.Logger) {
...
+
+	return nil
+}
+
+// isV2DatabaseSafeToDelete opens the v2 database read-only and checks whether
+// it contains any user data. Returns true only when all user-facing tables are
+// empty and no migration is in progress.
+func isV2DatabaseSafeToDelete(dbPath string, log logger.Logger) bool {
+	// Open database in read-only mode.
+	dsn := dbPath + "?mode=ro&_journal_mode=WAL&_busy_timeout=5000"
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/analysis/database_service.go:414]: SQLite DSN format is not a standard URL — it's a driver-specific connection string with pragma parameters. Using `net/url` would add complexity without benefit since this is a read-only open with hardcoded parameters, not user input.
